Maple2.Server.Game/PacketHandlers/MasteryHandler.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Maple2.Database.Storage;
+using Maple2.Model.Enum;
+using Maple2.Model.Error;
+using Maple2.Model.Game;
+using Maple2.Model.Metadata;
+using Maple2.PacketLib.Tools;
+using Maple2.Server.Core.Constants;
+using Maple2.Server.Core.PacketHandlers;
+using Maple2.Server.Game.Packets;
+using Maple2.Server.Game.Session;
+
+namespace Maple2.Server.Game.PacketHandlers;
+
+public class MasteryHandler : PacketHandler<GameSession> {
+    public override RecvOp OpCode => RecvOp.Mastery;
+
+    private enum Command : byte {
+        Reward = 1,
+        Craft = 2,
+    }
+
+    #region Autofac Autowired
+    // ReSharper disable MemberCanBePrivate.Global
+    public ItemMetadataStorage ItemMetadata { get; init; } = null!;
+    public TableMetadataStorage TableMetadata { private get; init; } = null!;
+    // ReSharper restore All
+    #endregion
+
+    public override void Handle(GameSession session, IByteReader packet) {
+        var command = packet.Read<Command>();
+        switch (command) {
+            case Command.Craft:
+                HandleCraft(session, packet);
+                break;
+        }
+    }
+
+    private void HandleCraft(GameSession session, IByteReader packet) {
+        int recipeId = packet.ReadInt();
+
+        if (!TableMetadata.MasteryRecipeTable.Entries.TryGetValue(recipeId, out MasteryRecipeTable.Entry? entry)) {
+            return;
+        }
+
+        // TODO: Check if player has completed the required quests.
+
+        if (session.Mastery[entry.Type] < entry.RequiredMastery) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_mastery));
+        }
+
+        if (session.Currency.CanAddMeso(-entry.RequiredMeso) != -entry.RequiredMeso) {
+            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_meso));
+            return;
+        }
+
+        var itemDictionary = new Dictionary<int, IEnumerable<Item>>();
```

Comment:
Rewrite using the same pattern I used for gemstones. (I did not test this so I don't even know if it compiles, but you get the idea)

```cs
// Build this index so we don't need to find materials twice.
Dictionary<int, IList<Item>> materialsById = entry.RequiredItems.ToDictionary(
    entry => entry.ItemId,
    entry => session.Item.Inventory.Find(entry.ItemId, entry.Rarity).ToList()
);
Dictionary<ItemTag, IList<Item>> materialsByTag = entry.RequiredItems.ToDictionary(
    entry => entry.Tag,
    entry => session.Item.Inventory.Filter(item => item.Metadata.Property.Tag == entry.Tag)
);

lock (session.Item) {
    foreach (MasteryRecipeTable.Ingredient ingredient in entry.RequiredItems) {
        int remaining = ingredient.Amount;
        if (ingredient.Tag != ItemTag.None) {
            foreach (Item material in materialsByTag[ingredient.Tag]) {
                remaining -= material.Amount;
                if (remaining <= 0) {
                    break;
                }
            }
        } else {
            foreach (Item material in materialsById[ingredient.ItemId]) {
                remaining -= material.Amount;
                if (remaining <= 0) {
                    break;
                }
            }
        }

        if (remaining > 0) {
            session.Send(MasteryPacket.Error(MasteryError.s_mastery_error_lack_item));
            return;
        }
    }

    foreach (MasteryRecipeTable.Ingredient ingredient in entry.RequiredItems) {
        int remaining = ingredient.Amount;
        if (ingredient.Tag != ItemTag.None) {
            foreach (Item material in materialsByTag[ingredient.Tag]) {
                int consume = Math.Min(remaining, material.Amount);
                if (!session.Item.Inventory.Consume(material.Uid, consume)) {
                    Logger.Fatal("Failed to consume item {ItemUid}", material.Uid);
                    throw new InvalidOperationException($"Fatal: Consuming item: {material.Uid}");
                }
    
                remaining -= consume;
                if (remaining <= 0) {
                    break;
                }
            }
        } else {
            foreach (Item material in materialsById[ingredient.ItemId]) {
                int consume = Math.Min(remaining, material.Amount);
                if (!session.Item.Inventory.Consume(material.Uid, consume)) {
                    Logger.Fatal("Failed to consume item {ItemUid}", material.Uid);
                    throw new InvalidOperationException($"Fatal: Consuming item: {material.Uid}");
                }
    
                remaining -= consume;
                if (remaining <= 0) {
                    break;
                }
            }
        }
    }
}
```

Reply:
Wouldn't this count ingredients twice if there is a matching item id and tag?
I think I get it tho. let me see what I can do

Reply:
It would ignore item id and only use tag.